src/mscorlib/shared/System/HashCode.cs

```diff
@@ -0,0 +1,433 @@
+            hash *= Prime2;
+            hash ^= hash >> 13;
+            hash *= Prime3;
+            hash ^= hash >> 16;
+            return hash;
+        }
+
+        public void Add<T>(T value)
+        {
+            Add(value?.GetHashCode() ?? 0);
+        }
+
+        public void Add<T>(T value, IEqualityComparer<T> comparer)
+        {
+            if (comparer is null)
+            {
+                Add(value);
+            }
+            else
+            {
+                Add(comparer.GetHashCode(value));
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void Add(int value)
+        {
+            // xxHash works as follows:
+            // 0. Initialize immediately. We can't do this in a struct (no default
+            //    ctor).
+            // 1. Accumulate blocks of length 16 (4 uints) into 4 accumulators.
+            // 2. Accumulate remaining blocks of length 4 (1 uint) into the hash.
+            // 3. Accumulate remaining blocks of length 1 into the hash.
+
+            // There is no need for *3 as this type only accepts ints. _queue1,
+            // _queue2 and _queue3 are basically a buffer so that when ToHashCode is
+            // called we can execute *2 correctly. That's what first three case
+            // statements do.
+
+            // We need to initialize the xxHash32 state (_v1 -> _v4) lazily (see *0)
+            // and the last place that can be done if you look at the original code
+            // is just before the first block of 16 bytes is mixed in. The xxHash32
+            // state is never used for streams containing fewer than 16 bytes.
+
+            // A bloom filter is used to determine whether the default case statement
+            // has even been executed. To do that we check if the length is smaller
+            // than 4 (_length ^ position will be non-zero if this is the case). The
+            // case statement is for values larger than 2, so the check only succeeds
+            // on exactly 3.
+
+            // To see what's really going on here, have a look at the Combine methods.
+            
+            var val = (uint)value;
+            uint position = _length % 4;
+
+            // Switch can't be inlined.
+
+            if (position == 0)
+                _queue1 = val;
+            else if (position == 1)
+                _queue2 = val;
+            else if (position == 2)
+                _queue3 = val;
+            else // == 3
+            {
+                // length smaller than 4?
+                if ((_length ^ position) == 0)
+                    Initialize(out _v1, out _v2, out _v3, out _v4);
+
+                _v1 = Round(_v1, _queue1);
+                _v2 = Round(_v2, _queue2);
+                _v3 = Round(_v3, _queue3);
+                _v4 = Round(_v4, val);
+            }
+
+            // Throw for more than uint.MaxValue fields.
+            _length = checked(_length + 1);
```

Comment:
I don't think it's super necessary either, if you are hashing uint.MaxValue fields you probably have far bigger DOS issues than HashCode re-initializing state. The strange `(_length ^ position) == 0` check came from the original code, which was more robust without leaning on `checked`:

```c#
(_v1 | _v2 | _v3 | _v4 | (_length ^ position) == 0)
```

It further decreases the chances of re-initialization to probably some crazy probability (`_v1 | (_position...` would probably be good enough). The comment about the bloom filter is also erroneous for this reason (and should be updated).

Should I restore the original bloom filter, restore the shorter option or simply drop the `checked`?

Reply:
You should drop the `checked`. I am not familiar with the details of the hashing algorithm, so I will leave the rest to your discretion. However, if you decide not to include the `_v1 | _v2 | _v3 | _v4` then it should just be `_length == position`, not `(_length ^ position) == 0`.

Reply:
If ```checked``` is taking a meaningful amount of time, you could probably just try
```
if(_length == Uint32.Max) 
{
    throw new InvalidOperationException(); // Or some other appropriate exception
}
```
Hopefully branch prediction can make it incredibly cheap since it'll never get taken.

I'm not specifically concerned about a DoS, rather just that this could start yielding incorrect results at a certain point. We shouldn't silently start producing incorrect data since it's incredibly difficult to debug or recover from.

Reply:
@morganbr if we're throwing an exception, I assume we should expose the `length` so that users don't have try/catch?

Reply:
I don't have strong feelings either way on exposing it. I also don't know the exact CLS rules, but yeah we probably would have to make it an int or a long so that all languages can reference it. If it's a long, I don't mind skipping an overflow check since 2^63 bytes is impossibly large, but I don't know how much that would slow down this code.

Reply:
I do not think we should be throwing here.

I think this should uint, and we should let it overflow silently.